src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs

```diff
@@ -864,28 +1004,68 @@ private IEnumerable ReadItemList()
 
             for (int i = 0; i < count; i++)
             {
-                ITaskItem item = ReadItem();
+                ITaskItem item = ReadTaskItem();
                 list.Add(item);
             }
 
             return list;
         }
 
+        private string ReadString()
+        {
+            return binaryReader.ReadString();
+        }
+
         private string ReadOptionalString()
         {
-            if (ReadBoolean())
+            if (fileFormatVersion < 10)
             {
-                return ReadString();
+                if (ReadBoolean())
+                {
+                    return ReadString();
+                }
+                else
+                {
+                    return null;
+                }
             }
-            else
+
+            return ReadDeduplicatedString();
+        }
+
+        private string ReadDeduplicatedString()
+        {
+            if (fileFormatVersion < 10)
             {
-                return null;
+                return ReadString();
             }
+
+            int index = ReadInt32();
+            return GetStringFromRecord(index);
         }
 
-        private string ReadString()
+        private string GetStringFromRecord(int index)
         {
-            return binaryReader.ReadString();
+            if (index == 0)
+            {
+                return null;
+            }
+            else if (index == 1)
+            {
+                return string.Empty;
+            }
+
+            // we reserve numbers 2-9 for future use.
+            // the writer assigns 10 as the index of the first string
+            index -= BuildEventArgsWriter.StringStartIndex;
```

Comment:
Why do this manipulation here rather than just adding the empty slots so the logical and physical indices match?

Reply:
Adding the empty slots felt wasteful ;) Also 10 is a number I pulled out of my hat. I doubt we'll use these special slots in the future I'm always scared of painting myself in a corner. Maybe it's just FUD though.

Reply:
Are there some strings (other than null and empty) that are far more common than others? I'm wondering if it would reduce the size of the binlog to have a Huffman encoding where the first bit being 0 indicates "special string" and 1 is other, and special means one of the most common strings and involves fewer bits. Probably too complicated to be worth it anyway.

Reply:
Yeah, a bit too complicated. Remember that the entire thing is also compressed by GZipStream.

Reply:
The index is a signed integer so you could have treated all negative numbers as reserved and have the (non-negative) indices for real strings match logical vs. physical.

E.g. `-1` for `null`, `-2` for `String.Empty`

Reply:
Oh, I see, that would waste space because of the 7bit encoding, never mind!